Terminal history

hexa00 opened the pull request:
See it in action:
![terminal-history](https://user-images.githubusercontent.com/698076/33220778-e0f847c0-d118-11e7-8b7e-f605cc2a70b4.gif)

svenefftinge commented on packages/process/src/node/process-backend-module.ts:
```diff
@@ -40,4 +41,8 @@ export default new ContainerModule(bind => {
+
+    bind(MultiRingBuffer).toSelf().inTransientScope();
+    /* 1MB size, TODO should be a user preference. */
+    bind(MultiRingBufferOptions).toConstantValue({ size: 1048576 });
```
Instead of user pref, I think allowing an env would be good enough.

hexa00 replied:
Humm I'd rather use a preference.. Is that a problem considering prefs come from the frontend? 
I'll have to tink about it, but I think it can wait a bit until @epatpol PRs are in.

svenefftinge commented on packages/terminal/src/browser/terminal-widget.ts:
```diff
@@ -270,6 +284,16 @@ export class TerminalWidget extends BaseWidget {
+         * a refresh for example won't close it.  */
+        if (this.closeOnDispose === true && this.terminalId !== undefined) {
+            this.shellTerminalServer.close(this.terminalId);
+        }
```
did you check whether `dispose` is actually called when closing the window?

hexa00 replied:
Yes it is, I tested closing the window closes the backend terminal , refresh diposes of everyting but the backend term.

svenefftinge commented on packages/process/src/node/multi-ring-buffer.ts:
```diff
@@ -0,0 +1,307 @@
+export interface WrapedPosition { newPos: number, wrap: boolean }
+
+@injectable()
+export class MultiRingBuffer {
```
The `MultiRingBuffer` seems to be a bit involved but easy to write tests against. Could you add some?

hexa00 replied:
Yes tests are in multi-ring-buffer.spec.ts ? code coverage is at 98%
So you see any missing ?  
Github doesn't show the diff because it's too bid.. maybe you missed it ?

svenefftinge replied:
Great. I missed them indeed. :)

hexa00 commented:
@svenefftinge yes that's the plan, I'm not sure how to organise that yet howewer I'll sketch something up but at least for now history is preserves on refresh/reconnect of a workspace :)

hexa00 commented on packages/process/src/node/multi-ring-buffer.ts:
```diff
@@ -0,0 +1,307 @@
+}
+
+export interface WrapedPosition { newPos: number, wrap: boolean }
+
```
OK

hexa00 commented:
I updated the PR with the most recent master.

Also noteworthy the I changed the output property of the TerminalProcess to be the createOutputStream()

Otherwise it was too easy to call process.output multiple times not realising that doing so creates multiple independent readers.

hexa00 commented:
Note the appveyor failure is unrelated:

````
lerna ERR! execute   1) node-extension-server list with search:
lerna ERR! execute      Error: 521: Origin Down for https://api.npms.io/v2/search?q=keywords%3Atheia-extension%20scope%3Atheia
```

kittaakos commented on packages/process/src/node/process-backend-module.ts:
```diff
@@ -42,4 +43,8 @@ export default new ContainerModule(bind => {
         }
     );
+
+    bind(MultiRingBuffer).toSelf().inTransientScope();
```
Just out of curiosity, if the `MultiRingBuffer` is not a singleton, how can it give a "unique" ID for the readers when calling `#getReader()`? Could this be a problem, or it is just fine as is. Thank you!

hexa00 replied:
This is OK since there's 1 `MultiRingBuffer` per process, the ids are unique per `MultiRingBuffer` if ti was a singleton you would get all the terminals output into a single ring buffer which would give funny results :)

hannamtt commented:
how is this enabled on the current master?

marcdumais-work commented:
> how is this enabled on the current master?

I think it's enabled by default, without having to do anything. Before this old PR, the content of terminals would be lost when the frontend was restarted.

hannamtt commented:
im using master and every new lunch has a new terminal :(

marcdumais-work commented:
> im using master and every new lunch has a new terminal :(

Can you describe what you are doing in more details? Do you use the Electron or browser version of an application?

akosyakov commented:
When a server is stopped so its child processes including shells.

hannamtt commented:
> > im using master and every new lunch has a new terminal :(
> 
> Can you describe what you are doing in more details? Do you use the Electron or browser version of an application?

im using the browser version and i am not stopping the server just closing the browser

marcdumais-work commented:
> im using the browser version and i am not stopping the server just closing the browser

Maybe you've found a bug. Normally terminals should be restored when reconnecting to the backend. See here I had a few terminals opened and they were all restored correctly after reloading the frontend: 

![example-app-terminals-restored-after-browser-restarted](https://user-images.githubusercontent.com/25749063/85337305-371a4880-b4ae-11ea-86f7-bf7cb26a3fd7.gif)
